workers/unity/Packages/com.improbable.gdk.querybasedinteresthelper/InterestQuery.cs

```diff
@@ -0,0 +1,117 @@
+        ///     A Constraint object defining the constraints of the query.
+        /// </param>
+        /// <remarks>
+        ///     Returns the full snapshot result by default.
+        /// </remarks>
+        /// <returns>
+        ///     An InterestQuery object.
+        /// </returns>
+        public static InterestQuery Query(Constraint constraint)
+        {
+            var interest = new InterestQuery
+            {
+                query =
+                {
+                    Constraint = constraint.AsQueryConstraint(),
+                    FullSnapshotResult = true,
+                    ResultComponentId = new List<uint>()
+                }
+            };
+            return interest;
+        }
+
+        /// <summary>
+        ///     Sets the maximum frequency of the query.
+        /// </summary>
+        /// <param name="frequency">
+        ///     The maximum frequency to return query results.
+        /// </param>
+        /// <returns>
+        ///     An updated InterestQuery object.
+        /// </returns>
+        public InterestQuery WithMaxFrequencyHz(float frequency)
+        {
+            query.Frequency = frequency;
+            return this;
+        }
+
+        /// <summary>
+        ///     Defines what components to return in the query results.
+        /// </summary>
+        /// <param name="resultComponentId">
+        ///     First ID of a component to return from the query results.
+        /// </param>
+        /// <param name="resultComponentIds">
+        ///     Further IDs of components to return from the query results.
+        /// </param>
+        /// <remarks>
+        ///     At least one component ID must be provided.
+        /// </remarks>
+        /// <returns>
+        ///     An updated InterestQuery object.
+        /// </returns>
+        public InterestQuery FilterResults(uint resultComponentId, params uint[] resultComponentIds)
+        {
+            var resultIds = new List<uint>(resultComponentIds.Length + 1) { resultComponentId };
+            resultIds.AddRange(resultComponentIds);
+
+            return FilterResults(resultIds);
+        }
+
+        /// <summary>
+        ///     Defines what components to return in the query results.
+        /// </summary>
+        /// <param name="resultComponentIds">
+        ///     Set of IDs of components to return from the query results.
+        /// </param>
+        /// <remarks>
+        ///     At least one component ID must be provided. Query results are not filtered
+        ///     if resultComponentIds is empty.
+        /// </remarks>
+        /// <returns>
+        ///     An updated InterestQuery object.
+        /// </returns>
+        public InterestQuery FilterResults(IEnumerable<uint> resultComponentIds)
+        {
+            if (!resultComponentIds.Any())
+            {
+                Debug.LogWarning("At least one InterestQuery must be provided to filter a query's results.");
```

Comment:
I can't find the conversation atm, what was the logic for deciding when we throw exceptions and when we log warnings?

Reply:
We log warnings in InterestTemplate and InterestQuery because providing an IEnumerable with 0 elements can be safely continued with a no-op

However with Constraint, providing an IEnumerable with 0 elements into Any or All would mean that the constraint being set would be an empty list - which is equivalent to not having a constraint set

The exception throwing is done because an empty QueryConstraint is considered an invalid by the runtime, so we want to stop users making that mistake as early as possible